vendor/api/v1/stripe/views.py

```diff
@@ -0,0 +1,194 @@
+        paid_date = timezone.datetime.fromtimestamp(self.event.data.object.status_transitions.paid_at)
+        amount_paid = str(self.event.data.object.total)
+
+    # TODO nice to move this try/except blocks inside a generic function in the StripeBaseAPI class
+        try:
+            customer_profile = CustomerProfile.objects.get(site=site, user__email__iequals=stripe_customer_email)
+        except ObjectDoesNotExist:
+            logger.error(f"StripeSubscriptionInvoicePaid error: email: {stripe_customer_email} does not exits")
+            return HttpResponse(status=400)
+
+        if 'stripe_id' not in customer_profile.meta:
+            customer_profile.meta['stripe_id'] = stripe_customer_id
+            customer_profile.save()
+
+        try:
+            subscription = Subscription.objects.get(meta__stripe_id=stripe_subscription_id, profile=customer_profile)
+        except ObjectDoesNotExist:
+            logger.error(f"StripeSubscriptionInvoicePaid customer: {customer_profile} does not have a subscription with stripe_id: {stripe_subscription_id}")
+            return HttpResponse(status=400)
+
+        offer = subscription.get_offer()
+
+        if not offer:
+            logger.error(f"StripeSubscriptionInvoicePaid {subscription} has no offer attached")
+            return HttpResponse(status=400)
+
+        invoice = Invoice.objects.create(
+            profile=customer_profile,
+            site=site,
+            ordered_date=paid_date,
+            total=float(f"{amount_paid[:-2]}.{amount_paid[-2:]}"),
+            status=InvoiceStatus.COMPLETE
+        )
+        invoice.add_offer(offer)
+        invoice.vendor_notes = {'stripe_id': stripe_invoice_id}
+        invoice.save()
+
+        processor = StripeProcessor(site, invoice)
+        processor.renew_subscription(subscription, stripe_transaction_id, PurchaseStatus.SETTLED, payment_success=True)
+
+        return HttpResponse(status=200)
+
+class StripeSubscriptionPaymentFailed(StripeBaseAPI):
+
+    def post(self, request, *args, **kwargs):
+        site = get_site_from_request(request)
+
+        if not self.is_valid_post(site):
+            return HttpResponse(status=400)
+
+        if not self.is_incoming_event_correct(self.event, StripeEvents.INVOICE_PAYMENT_FAILED):
+            return HttpResponse(status=400)
+
+        # TODO nice to move this assignments inside a generic function in the StripeBaseAPI class
+        stripe_subscription_id = self.event.data.object.subscription
+        stripe_customer_id = self.event.data.object.customer
+        stripe_customer_email = self.event.data.object.customer_email
+        stripe_transaction_id = self.event.data.object.charge
+        stripe_invoice_id = self.event.data.object.stripe_id
+        paid_date = timezone.datetime.fromtimestamp(self.event.data.object.status_transitions.paid_at)
+
+        # TODO nice to move this try/except blocks inside a generic function in the StripeBaseAPI class
+        try:
```

Comment:
i agree- looks like you could create a function to reuse a lot of the logic between the successful and failed payments views :) later, when you have some time